File: mobile/test/receipt_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:json_schema/json_schema.dart';
import 'package:kanean/contract/receipt_contract.dart';

/// 契約のゴールデンテスト（design D4）。
///
/// **TS 側の `packages/shared/src/__tests__/receiptContract.test.ts` と同じファイルを読む。**
/// 片側だけ解釈を変えるとどちらかが落ちる — これがドリフトに対する唯一の防壁。
/// 参照するのは zod から生成された JSON Schema そのもので、Dart 側の写しは作らない。
void main() {
  final contractDir = Directory('../packages/shared/contract/receipt');
  final fixturesDir = Directory('${contractDir.path}/fixtures');

  Map<String, dynamic> readJson(String path) =>
      jsonDecode(File(path).readAsStringSync()) as Map<String, dynamic>;

  final manifest = readJson('${fixturesDir.path}/index.json');
  final cases = (manifest['cases'] as List<dynamic>).cast<Map<String, dynamic>>();

  final schemas = {
    'meta': JsonSchema.create(readJson('${contractDir.path}/receipt-meta.schema.json')),
    'status': JsonSchema.create(readJson('${contractDir.path}/receipt-status.schema.json')),
  };

  group('レシート契約のゴールデン', () {
    test('契約の成果物が置かれている', () {
      expect(contractDir.existsSync(), isTrue,
          reason: 'packages/shared の build:schema を回していない可能性がある');
      expect(cases, isNotEmpty);
    });

    for (final c in cases) {
      final file = c['file'] as String;
      final kind = c['kind'] as String;
      final valid = c['valid'] as bool;

      test('$file は ${valid ? '受理される' : '弾かれる'}', () {
        final json = readJson('${fixturesDir.path}/$file');
        // まず zod 由来の JSON Schema。TS 側と同じ判定になることを確かめる。
        expect(schemas[kind]!.validate(json).isValid, valid);

        if (!valid) return;
        // 次に Dart のモデル。スキーマが通るものは必ず型として読める。
        if (kind == 'meta') {
          final meta = ReceiptMeta.fromJson(json);
          expect(meta.schemaVersion, receiptSchemaVersion);
          // 往復して元の JSON と同じものに戻る＝取りこぼした項目が無い。
          expect(meta.toJson(), equals(json));
        } else {
          final status = ReceiptStatus.fromJson(json);
          expect(status.schemaVersion, receiptSchemaVersion);
          expect(status.toJson(), equals(json));
        }
      });
    }

    test('現金とカードを取り違えない', () {
      final cash = ReceiptMeta.fromJson(readJson('${fixturesDir.path}/meta-cash-full.json'));
      final card = ReceiptMeta.fromJson(readJson('${fixturesDir.path}/meta-card.json'));
      expect(cash.paymentMethod, PaymentMethod.cash);
      expect(card.paymentMethod, PaymentMethod.card);
    });

    test('未登録の status は必ず理由を持つ', () {
      for (final c in cases.where((c) => c['kind'] == 'status' && c['valid'] == true)) {
        final status = ReceiptStatus.fromJson(readJson('${fixturesDir.path}/${c['file']}'));
        if (!status.registered) expect(status.reason, isNotNull);
      }
    });

    test('登録済みの status は帳簿の中身を運ばない', () {
      final json = readJson('${fixturesDir.path}/status-registered.json');
      // 要約に許すのは entryId・date・totalAmount・accountName の4つだけ。
      final keys = (json['summary'] as Map<String, dynamic>).keys.toList()..sort();
      expect(keys, ['accountName', 'date', 'entryId', 'totalAmount']);
    });

    test('読み取れなかった項目は空のまま運べる', () {
      // 端末 OCR が読めなくても撮影は成立する（receipt-capture spec）。
      final meta = ReceiptMeta.fromJson(readJson('${fixturesDir.path}/meta-cash-minimal.json'));
      expect(meta.ocr, isNull);
      expect(meta.usage, isNull);
      expect(meta.meal, isNull);
      // 支払手段だけは欠けない。
      expect(meta.paymentMethod, PaymentMethod.cash);
    });
  });
}

File: packages/shared/src/__tests__/receiptContract.test.ts
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import { z } from 'zod'
import { receiptMetaSchema, receiptStatusSchema } from '../receipt.js'

/**
 * 契約のゴールデンテスト（design D4）。**Dart 側のテストが同じフィクスチャを読む**ので、
 * 片側だけ解釈を変えるとどちらかが落ちる — これがドリフトに対する唯一の防壁。
 * 契約を変えるときはフィクスチャを先に足すこと。
 */

const contractDir = path.join(
  path.dirname(fileURLToPath(import.meta.url)),
  '..',
  '..',
  'contract',
  'receipt',
)
const fixturesDir = path.join(contractDir, 'fixtures')

interface Case {
  file: string
  kind: 'meta' | 'status'
  valid: boolean
}

const readJson = (p: string): unknown => JSON.parse(readFileSync(p, 'utf8'))
const manifest = readJson(path.join(fixturesDir, 'index.json')) as { cases: Case[] }
const schemas = { meta: receiptMetaSchema, status: receiptStatusSchema }

describe('レシート契約のゴールデン', () => {
  it('フィクスチャの一覧が空でない', () => {
    expect(manifest.cases.length).toBeGreaterThan(0)
  })

  for (const c of manifest.cases) {
    it(`${c.file} は ${c.valid ? '受理される' : '弾かれる'}`, () => {
      const result = schemas[c.kind].safeParse(readJson(path.join(fixturesDir, c.file)))
      expect(result.success).toBe(c.valid)
    })
  }

  it('現金と カードを取り違えない', () => {
    const cash = receiptMetaSchema.parse(readJson(path.join(fixturesDir, 'meta-cash-full.json')))
    const card = receiptMetaSchema.parse(readJson(path.join(fixturesDir, 'meta-card.json')))
    expect(cash.paymentMethod).toBe('cash')
    expect(card.paymentMethod).toBe('card')
  })

  it('未登録の status は必ず理由を持つ', () => {
    for (const c of manifest.cases.filter((x) => x.kind === 'status' && x.valid)) {
      const status = receiptStatusSchema.parse(readJson(path.join(fixturesDir, c.file)))
      if (status.outcome === 'skipped') expect(status.reason).toBeTruthy()
    }
  })

  it('登録済みの status は帳簿の中身を運ばない', () => {
    const status = receiptStatusSchema.parse(
      readJson(path.join(fixturesDir, 'status-registered.json')),
    )
    // 要約に許すのは entryId・date・totalAmount・accountName の4つだけ（receipt-inbox spec）。
    expect(status.outcome).toBe('registered')
    if (status.outcome !== 'registered') return
    expect(Object.keys(status.summary).sort()).toEqual([
      'accountName',
      'date',
      'entryId',
      'totalAmount',
    ])
  })
})

describe('生成された JSON Schema', () => {
  // Dart 側は TS を実行できないので、追跡している生成物が唯一の入力になる。
  // zod を直して再生成し忘れると Dart だけ古い契約で動くため、ここで検出する。
  const targets = [
    { file: 'receipt-meta.schema.json', name: 'ReceiptMeta', schema: receiptMetaSchema },
    { file: 'receipt-status.schema.json', name: 'ReceiptStatus', schema: receiptStatusSchema },
  ]

  for (const { file, name, schema } of targets) {
    it(`${file} が zod と一致している（ズレたら build:schema を回す）`, () => {
      const { $schema, ...body } = z.toJSONSchema(schema, { target: 'draft-7' })
      const generated = { $ref: `#/definitions/${name}`, definitions: { [name]: body }, $schema }
      expect(readJson(path.join(contractDir, file))).toEqual(generated)
    })
  }
})
